Config policy for self-hosted runner how to (#7787)
circleci/circleci-docs: jekyll/_cci2/config-policies-for-self-hosted-runner.adoc

contentTags:
  platform:
  - Cloud
---
= Config policies for self-hosted runner (Open-preview)
:page-layout: classic-docs
:page-liquid:
:icons: font
:toc: macro
:toc-title:

NOTE: The config policies feature is available on the **Scale** plan and is currently in **open preview**.

CAUTION: While the config policies feature is in **open preview**, a service failure will result in build configurations **not** being evaluated against policies. This should be taken into consideration before using config policies for compliance purposes.

Follow this how-to guide to learn how to create a config policy to restrict which projects can run jobs on a self-hosted runner resource class. For more information about config policies, see the xref:config-policy-management-overview#[Config policies overview].

[#prerequisites]
== Prerequisites

* Install/update the CircleCI CLI, and ensure you have authenticated with a token before attempting to use the CLI with config policies. See the xref:local-cli#[Installing the Local CLI] page for more information.

* Set up a self-hosted runner. See the xref:runner-installation#[Installing self-hosted runners] guide.

> * Ensure you have **enabled** config policy evaluation for your organization so that project configurations **will** be evaluated against your organization's policies when pipelines are triggered.:

Author: ```suggestion
* Ensure you have **enabled** config policy evaluation for your organization so that project configurations **will** be evaluated against your organization's policies when pipelines are triggered:
```

> * Ensure you have **enabled** config policy evaluation for your organization so that project configurations **will** be evaluated against your organization's policies when pipelines are triggered.:

Author: 🤣

+
[source,shell]
----
circleci policy settings --enabled=true --owner-id <your-organization-ID>
----
+
Example output:
+
[source,shell]
----
{
  "enabled": true
}
---- 
+
{% include snippets/find-organization-id.adoc %}

[#create-your-policy]
== 1. Create your policy

. If you have not already done so, create an empty directory to store your policies. For example:
+
[source,shell]
----
mkdir ./config-policies
----

> . Create a new file in this directory for your new policy. Name the file `runner.rego`

Reviewer: There should be a mention towards managing policies via VCS, we don't want users to be creating policies from many sources as policies need one centralized source

Author: OK @davidmdm I've made a separate PR to move the management through VCS steps to be part of the initial setup and recommendations, rather than just being presented as an "option" https://github.com/circleci/circleci-docs/pull/7829/files. I'll handle this change separately to this policies for runner PR if that works

> . Create a new file in this directory for your new policy. Name the file `runner.rego`

Reviewer 2: ```suggestion
. Create a new file in this directory for your new policy. Name the file `runner.rego`.
```
Nit

. Copy the following snippet into the `runner.rego` file you just made:
+
[source,rego]
----
package org
import data.circleci.config

policy_name["runner_policies"]

enable_rule["resource_class_check"]

hard_fail["resource_class_check"]

resource_class_check = config.resource_class_by_project({ "namespace/resource_class": {"project_UUIDs"}})
----

> +

Reviewer: it can be mentioned that it can be multiple project_uuids

Author: This is shown in the "more complex example" at the end of the page, but maybe I should highlight this throughout?

> In the following steps you will replace `namespace/resource_class` and `project_UUIDs` with one of your self-hosted runner resource classes, and one of your existing projects. The `runner.rego` policy, once uploaded, will then restrict your specified self-hosted runner to only run jobs from your specified project.

Author: ```suggestion
In the following steps you will replace `namespace/resource_class` and `project_UUIDs` with one of your self-hosted runner resource classes, and one of your existing projects (you can also add multiple projects). The `runner.rego` policy, once uploaded, will then restrict your specified self-hosted runner to only run jobs from your specified project.
```

.. Replace `namespace/resource_class` in the `runner.rego` file with the name of the resource class that you would like to restrict.

. Replace `project_UUIDs` with a project ID:
.. In the CircleCI web app, navigate to your projects dashboard by selecting **Projects** in the sidebar. Find the project you want to allow to build on your self-hosted runner, and then click the ellipsis (`...`) next to that project and select **Project Settings**. 
.. Copy the Project ID from the overview page and replace `project_UUIDs` with this project ID.

[#push-up-your-policy-bundle]
== 3. Push up your policy bundle

Create and upload the policy bundle using CircleCI CLI:

[source,shell]
----
circleci policy push ./config-policies –owner-id <your-organization-ID>
----

If the upload was successful, you will see something like the following:

[source,shell]
----
{
  “Created”: [“runner_policies”]
}
----

> NOTE: If you would like to write tests for your policy, check out the link:/docs/test-config-policies/[Test config policies] guide.

Reviewer 2: ```suggestion
NOTE: If you would like to write tests for your policy, check out the xref:test-config-policies#[Test config policies] guide.
```

[#a-more-complex-example]
== A more complex example

Once you have this runner restriction policy up and running, you can quickly add more runner resource class/project pairings, and allow multiple projects per runner. To do this you will expand the `resource_class_check` rule.

The following snippet shows two separate runner resource classes (`my-namespace/runner-test1` and `my-namespace/runner-test2`) with allowed projects, the first has two projects in its allowed list.

[source,rego]
----

package org
import data.circleci.config

policy_name["runner_policies"]

enable_rule["resource_class_check"]

hard_fail["resource_class_check"]

resource_class_check = config.resource_class_by_project({ 
  "my-namespace/runner-test1": {
    "a20ae1c6-d723-4c03-9bb6-01d5b3594e02",
    "2accadd0-c832-489e-b837-4de03def0d35"
  },
  "my-namespace/runner-test2": {
    "637f9136-694e-4105-bb43-1028ffb9043e"
  }
})

----

[#next-steps]
== Next steps

> * link:/docs/create-and-manage-config-policies[Create and manage config policies]

Reviewer 2: ```suggestion
* xref:create-and-manage-config-policies#[Create and manage config policies]
```

> * link:/docs/test-config-policies/[Test config policies]

Reviewer 2: ```suggestion
* xref:test-config-policies#[Test config policies]
```

> * link:/docs/config-policy-reference[Config policy reference]

Reviewer 2: ```suggestion
* xref:config-policy-reference#[Config policy reference]
```
